staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/pruning/algorithm.go

```diff
@@ -43,70 +69,105 @@ var metaFields = map[string]bool{
 	"metadata":   true,
 }
 
-func prune(x interface{}, s *structuralschema.Structural) {
+func appendKey(path []string, key string) []string {
+	if len(path) > 0 {
+		path = append(path, ".")
+	}
+	return append(path, key)
+}
+
+func appendIndex(path []string, index int) []string {
+	return append(path, "[", strconv.Itoa(index), "]")
+}
+
+func prune(x interface{}, s *structuralschema.Structural, opts PruneOptions) []string {
 	if s != nil && s.XPreserveUnknownFields {
-		skipPrune(x, s)
-		return
+		return skipPrune(x, s, PruneOptions{
+			parentPath: opts.parentPath,
+		})
 	}
 
+	var pruned []string
+	origPathLen := len(opts.parentPath)
 	switch x := x.(type) {
 	case map[string]interface{}:
 		if s == nil {
 			for k := range x {
+				if !metaFields[k] {
+					pruned = append(pruned, strings.Join(appendKey(opts.parentPath, k), ""))
+				}
 				delete(x, k)
+				opts.parentPath = opts.parentPath[:origPathLen]
 			}
-			return
+			return pruned
 		}
 		for k, v := range x {
 			if s.XEmbeddedResource && metaFields[k] {
 				continue
 			}
+			opts.parentPath = appendKey(opts.parentPath, k)
```

Comment:
Sorry, it's late and my brain is fogging on my. I'm not sure I understand how we are not retaining  increases in allocation.  First we append and that grows the slice, then we do `parentPath[:len]` to shorten it. Does this actually reduce the capcity? Is there a way to do this without reducing the capacity? Or are you saying this something specific about this `appendKey` helper that is allocating new slices when it shouldn't?

Reply:
Because `opts` is not pass-by-reference, when you assign to `opts.parentPath`, you're not actually propagating a newly-allocated-with-more-capacity slice back up the call tree so the new extra capacity can be used by parent/peer traversals.

This isn't a correctness issue, but it is an efficiency issue. Right now, this code is exercised all the time, even if the feature is off. Possibilities to unblock this PR:
1. benchmark and reduce allocations here by propagating newly allocated bigger parentPath slices upward to callers
2. make the parentPath tracking inert if the feature gate is off, and add open an issue tracking benchmarking and reducing allocations here before beta